geoportailv3/static/js/search/searchdirective.js

```diff
@@ -0,0 +1,448 @@
+        if (!goog.object.containsKey(suggestion, 'bglayer')) {
+          scope['click'] = goog.bind(function(event) {
+            this.showLayerinfo_(this.getLayerFunc_(suggestion));
+            event.stopPropagation();
+          }, this);
+          html += '<button ng-click="click($event)">i</button>';
+        } else {
+          html += ' (' + gettextCatalog.getString('Background') + ') ';
+        }
+        html += '</p>';
+        return $compile(html)(scope);
+      }, this)
+    })
+  }
+  ];
+
+  this['listeners'] = /** @type {ngeox.SearchDirectiveListeners} */ ({
+    selected: goog.bind(app.SearchDirectiveController.selected_, this)
+  });
+
+};
+
+
+/**
+ * @return {ol.FeatureOverlay} The feature overlay.
+ * @private
+ */
+app.SearchDirectiveController.prototype.createFeatureOverlay_ = function() {
+  var featureFill = new ol.style.Fill({
+    color: [255, 255, 0, 0.6]
+  });
+  var featureStroke = new ol.style.Stroke({
+    color: [255, 155, 55, 1],
+    width: 3
+  });
+  var featureOverlay = new ol.FeatureOverlay({
+    style: new ol.style.Style({
+      fill: featureFill,
+      stroke: featureStroke,
+      image: new ol.style.Circle({
+        radius: 10,
+        fill: featureFill,
+        stroke: featureStroke
+      })
+    })
+  });
+  featureOverlay.setMap(this['map']);
+  return featureOverlay;
+};
+
+
+/**
+ * @param {ngeo.CreateGeoJSONBloodhound} ngeoCreateGeoJSONBloodhound The ngeo
+ *     create GeoJSON Bloodhound service.
+ * @param {string} searchServiceUrl
+ * @return {Bloodhound} The bloodhound engine.
+ * @private
+ */
+app.SearchDirectiveController.prototype.createAndInitPOIBloodhound_ =
+    function(ngeoCreateGeoJSONBloodhound, searchServiceUrl) {
+  var url = searchServiceUrl + '?limit=5&query=%QUERY';
+  var bloodhound = ngeoCreateGeoJSONBloodhound(url, ol.proj.get('EPSG:3857'));
+  bloodhound.initialize();
+  return bloodhound;
+};
+
+
+/**
+ * @return {Bloodhound} The bloodhound engine.
+ * @private
+ */
+app.SearchDirectiveController.prototype.createAndInitLayerBloodhound_ =
+    function() {
+  var bloodhound = new Bloodhound({
+    datumTokenizer: Bloodhound.tokenizers.obj.whitespace('translatedName'),
+    queryTokenizer: Bloodhound.tokenizers.whitespace,
+    local: []
+  });
+  bloodhound.initialize();
+  return bloodhound;
+};
+
+
+/**
+ * @param {app.Themes} appThemes Themes Service
+ * @param {Bloodhound} bloodhound
+ * @param {angularGettext.Catalog} gettextCatalog Gettext catalog
+ * @private
+ */
+app.SearchDirectiveController.prototype.createLocalAllLayerData_ =
+    function(appThemes, bloodhound, gettextCatalog) {
+  bloodhound.clear();
+  this.layers_ = [];
+  appThemes.getBgLayers().then(
+      goog.bind(function(bgLayers) {
+        goog.array.forEach(bgLayers, function(bgLayer) {
+          var bgLayerObj = {};
+          bgLayerObj.bglayer = bgLayer;
+          bgLayerObj.translatedName =
+              gettextCatalog.getString(bgLayer.values_.label);
```

Comment:
You use the dot notation here for `bglayer` and `translatedName`, meaning that you allow the compiler to rename these properties. But elsewhere you refer to these properties using strings.

For example you use `displayKey: 'translatedName'` and `tokenizers.obj.whitespace('translatedName')`. And you use `goog.object.containsKey(input, 'bglayer')`.

This is not correct. You need to be consistent in the way you use object properties. What I'd do is create a typedef for the layer suggestion objects and consistently use the dot notation to refer to the properties.

```js
/**
 * @typedef {{bgLayer: ol.layer.Layer, translatedName: string}}
 */
app.LayerSuggestion;
```

and then I'd use code like this:

```js
    var suggestions = goog.array.map(bgLayers,
        /**
         * @param {ol.layer.Layer} bgLayer
         * @return {app.LayerSuggestion} Layer suggestion.
         */
         function(bgLayer) {
           return {bgLayer: bgLayer,
               translatedName: gettextCatalog.getString(bgLayer.get('label')};
         });
```